### PR TITLE
FIBONACCI_SEARCH in cpp

### LINKED ISSUE #354: Fibonacci Search
Adding Fibonacci Search under Search Section in java
Being a part of the GSSOC'21, I would like to implement this in java .Please do assign this to me

### PR DESCRIPTION
**Pull request by Op-panda**
Added **FIBONACCI SEARCH** algorithm with proper explaination and comments in CPP
please check this out      issue #354

### CODE REVIEW

**Comment by Op-panda** [comment]
@tarun26091999 please review this.

**Review by raksha009 — COMMENTED** [review]
Plz link your code with appropriate folder, Go through the guidelines and the pinned issues for the same

File: Algorithm/Searching_Sorting/fibonacci_search.cpp [hunk 1]
```diff
@@ -0,0 +1,161 @@
+/*
+Fibonacci search is an efficient search algorithm based on:
+--> divide and conquer principle that can find an element in the given sorted array. */
+/*
```
Comment (raksha009) [inline Algorithm/Searching_Sorting/fibonacci_search.cpp:4]: Include the entire description within 1 multiline comment.

File: Algorithm/Searching_Sorting/fibonacci_search.cpp [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+{
+    // a,b,c are variables that stores the fibonacci numbers sequentially.
+    int a = 0;
+    int b = 1;
+    int c = a + b;
+
+    // until c does not become equal to or greater than n , loop executes.
+    while (c < n)
+    {
+
```
Comment (raksha009) [inline Algorithm/Searching_Sorting/fibonacci_search.cpp:63]: Remove this extra line

File: Algorithm/Searching_Sorting/fibonacci_search.cpp [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    // checking if c is at valid location.
+    while (c > 1)
+    {
+        // i will be assigned the value of min() used.
+        int i = min(offset + a, n - 1);
+
+        /*If x is greater than the value at index ,c cut the subarray array from offset to i*/
+        if (arr[i] < x)
+        {
+
```
Comment (raksha009) [inline Algorithm/Searching_Sorting/fibonacci_search.cpp:80]: Remove this extra line

File: Algorithm/Searching_Sorting/fibonacci_search.cpp [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        /*If x is greater than the value at index ,c
+cut the subarray after i+1. */
+        else if (arr[i] > x)
+        {
+            c = a;
+            b = b - a;
+            a = c - b;
+        }
+
+        else
```
Comment (raksha009) [inline Algorithm/Searching_Sorting/fibonacci_search.cpp:95]: Remove this extra line

File: Algorithm/Searching_Sorting/fibonacci_search.cpp [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+            //if element found, return index.
+            return i;
+    }
+    //comparing the last element with x
```
Comment (raksha009) [inline Algorithm/Searching_Sorting/fibonacci_search.cpp:99]: Add a line here

File: Algorithm/Searching_Sorting/fibonacci_search.cpp [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+int main()
+{
+    int t;
+    cin >> t;
+    while (t--)
+    {
+        int l;
+        cout << "\nEnter the number of elements in array which should be less than 10";
+        cin >> l;
+        //array defined
```
Comment (raksha009) [inline Algorithm/Searching_Sorting/fibonacci_search.cpp:115]: Remove this comment

File: Algorithm/Searching_Sorting/fibonacci_search.cpp [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+}
+
+/* complexities */
+
+/* Worst case time complexity:  O(log n)
+Average case time complexity:  O(log n)
+Best case time complexity:  O(1)
+Space complexity:  O(1)   */
+
+/* sample input :
```
Comment (raksha009) [inline Algorithm/Searching_Sorting/fibonacci_search.cpp:141]: Follow 
Test Cases:
Input 
Output

Include all the part below code in single multiline comment

**Review by raksha009 — CHANGES_REQUESTED** [review]

File: Algorithm/Searching_Sorting/fibonacci_search.cpp (hunk 4 above)
Comment (Op-panda) [reply Algorithm/Searching_Sorting/fibonacci_search.cpp:90]: changes done

File: Algorithm/Searching_Sorting/fibonacci_search.cpp (hunk 5 above)
Comment (Op-panda) [reply Algorithm/Searching_Sorting/fibonacci_search.cpp:95]: changes done

File: Algorithm/Searching_Sorting/fibonacci_search.cpp [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+    // Marks the eliminated range from front.
+    int offset = -1;
+
+    // checking if c is at valid location.
+    while (c > 1)
+    {
+        // i will be assigned the value of min() used.
+        int i = min(offset + a, n - 1);
+
+        /*If x is greater than the value at index ,c cut the subarray array from offset to i*/
```
Comment (raksha009) [inline Algorithm/Searching_Sorting/fibonacci_search.cpp:74]: Do not use multiline comments in between the code. Use single line comment here

File: Algorithm/Searching_Sorting/fibonacci_search.cpp [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+        if (arr[i] < x)
+        {
+            c = b;
+            b = a;
+            a = c - b;
+            offset = i;
+        }
+        /*If x is greater than the value at index ,c
```
Comment (raksha009) [inline Algorithm/Searching_Sorting/fibonacci_search.cpp:82]: Do not use multiline comments in between the code. Use single line comment here

File: Algorithm/Searching_Sorting/fibonacci_search.cpp [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+cut the subarray after i+1. */
```
Comment (raksha009) [inline Algorithm/Searching_Sorting/fibonacci_search.cpp:83]: combine it with above line.

File: Algorithm/Searching_Sorting/fibonacci_search.cpp [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+    //comparing the last element with x
+    if (b && arr[offset + 1] == x)
+        return offset + 1;
+    // if not then return -1.
+    return -1;
+}
+
+int main()
+{
+    int t;
```
Comment (raksha009) [inline Algorithm/Searching_Sorting/fibonacci_search.cpp:104]: Specify what is t

File: Algorithm/Searching_Sorting/fibonacci_search.cpp [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+        int x;
+        cout << "\nEnter element to be searched :";
+        cin >> x;
+        cout << "Found at index:" << FibonacciSearch(arr, x, n);
+    }
+    return 0;
+}
+
+/* 
+test cases:
```
Comment (raksha009) [inline Algorithm/Searching_Sorting/fibonacci_search.cpp:130]:
```suggestion
Test Cases:

```

File: Algorithm/Searching_Sorting/fibonacci_search.cpp [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+
+input :
```
Comment (raksha009) [inline Algorithm/Searching_Sorting/fibonacci_search.cpp:132]:
```suggestion
Input :

```

File: Algorithm/Searching_Sorting/fibonacci_search.cpp [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+100 90 30 15 60 120 10
+30
+5
+40 60 22 10 45
+22
+2
+40 60
+45
+
+output:
```
Comment (raksha009) [inline Algorithm/Searching_Sorting/fibonacci_search.cpp:144]:
```suggestion
Output:

```

File: Algorithm/Searching_Sorting/fibonacci_search.cpp [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+2
+1
+-1   */
```
Comment (raksha009) [inline Algorithm/Searching_Sorting/fibonacci_search.cpp:147]: Add 2 test cases and add Space and time complexities too

File: Algorithm/Searching_Sorting/readme.md [hunk 16]
```diff
@@ -57,6 +57,7 @@ Following are the types of searches which we will be discussing in this book.
 * Binary Search Trees 
 * Symbol Tables and Hashing
 * String Searching Algorithms: Tries, Ternary Search and Suffix Trees
+* Fibonacci Search:[c++](Algorithm/Searching_Sorting/fibonacci_search.cpp)
```
Comment (raksha009) [inline Algorithm/Searching_Sorting/readme.md:60]:
```suggestion
* Fibonacci Search:  [C++](Algorithm/Searching_Sorting/fibonacci_search.cpp)
```

Comment (raksha009) [inline Algorithm/Searching_Sorting/readme.md:60]: Add this line at line 70

**Review by raksha009 — CHANGES_REQUESTED** [review]

**Review by raksha009 — CHANGES_REQUESTED** [review]
Plz do the required changes

File: Algorithm/Searching_Sorting/fibonacci_search.cpp (hunk 11 above)
Comment (Op-panda) [reply Algorithm/Searching_Sorting/fibonacci_search.cpp:104]: done

File: Algorithm/Searching_Sorting/fibonacci_search.cpp (hunk 9 above)
Comment (Op-panda) [reply Algorithm/Searching_Sorting/fibonacci_search.cpp]: done

**Comment by Op-panda** [comment]
done all changes you asked , I hope this is fine now??

File: Algorithm/Searching_Sorting/fibonacci_search.cpp [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+1
+
+Input 3:
+2
+40 60
+45
+Output 3:
+-1
+
+Space complexity: 4
```
Comment (raksha009) [inline Algorithm/Searching_Sorting/fibonacci_search.cpp:148]:
```suggestion
Space complexity: O(1)

```

**Review by raksha009 — CHANGES_REQUESTED** [review]
Do this small there, then we are good to go

**Comment by raksha009** [comment]
Also, resolve the conflicts

**Comment by Op-panda** [comment]
Hey, @raksha009, changes done ,plss tell me if this is okay or some more changes to be done 
@rudrakshi99

**Review by raksha009 — APPROVED** [review]

**Pull request merged**